request-chroot_test.go

```diff
@@ -0,0 +1,428 @@
+}
+
+func TestChrootRemove(t *testing.T) {
+	rootPath, client := getChrootTuple(t)
+	defer client.Close(t)
+
+	const (
+		dirName  = "dir"
+		fileName = "file-to-remove"
+	)
+	// prepare
+	require.NoError(t, os.MkdirAll(filepath.Join(rootPath, dirName), 0755))
+	require.NoError(t, ioutil.WriteFile(filepath.Join(rootPath, fileName), []byte{}, 0644))
+	// file should be removed
+	assert.NoError(t, client.Remove(fileName))
+	assert.NoFileExists(t, filepath.Join(rootPath, fileName))
+	// remove non-exist file should fail
+	assert.Error(t, client.Remove(fileName))
+	// can remove a folder
+	assert.NoError(t, client.Remove(dirName))
+	AssertNoDirExist(t, filepath.Join(rootPath, dirName))
+
+	// outer path will be converted relate to root, and /{root}/outbound is not exist
+	assert.ErrorIs(t, client.Remove(outbound), os.ErrNotExist)
+}
+
+func TestChrootRename(t *testing.T) {
+	rootPath, client := getChrootTuple(t)
+	defer client.Close(t)
+
+	const (
+		oldfile = "oldfile"
+		newfile = "newfile"
+		extfile = "extfile"
+		olddir  = "olddir"
+		newdir  = "newdir"
+	)
+	require.NoError(t, os.MkdirAll(filepath.Join(rootPath, olddir), 0755))
+	require.NoError(t, ioutil.WriteFile(filepath.Join(rootPath, oldfile), []byte{}, 0644))
+	require.NoError(t, ioutil.WriteFile(filepath.Join(rootPath, extfile), []byte{}, 0644))
+	// call sftp cmd
+	assert.NoError(t, client.Rename(oldfile, newfile))
+	// check result, ori file should rename to new
+	assert.NoFileExists(t, filepath.Join(rootPath, oldfile))
+	assert.FileExists(t, filepath.Join(rootPath, newfile))
+	// unlike POSIX, SFTP forbid rename to an exist file
+	assert.Error(t, client.Rename(newfile, extfile))
+	// can rename a folder
+	assert.NoError(t, client.Rename(olddir, newdir))
+	AssertDirExist(t, filepath.Join(rootPath, newdir))
+
+	// we don't want client user escape the root
+	assert.Error(t, client.Rename(outbound, olddir))
+}
+
+func TestChrootSymlink(t *testing.T) {
+	rootPath, client := getChrootTuple(t)
+	defer client.Close(t)
+
+	const (
+		regular = "/path/regular"
+		symlink = "symlink"
+	)
+	// prepare
+	content := []byte(strings.Repeat("hello sftp", 1024))
+	require.NoError(t, os.MkdirAll(filepath.Dir(filepath.Join(rootPath, regular)), 0755))
+	require.NoError(t, ioutil.WriteFile(filepath.Join(rootPath, regular), content, 0644))
+	// call sftp cmd
+	assert.NoError(t, client.Symlink(regular, symlink))
+	// symlink should be created
+	f, err := os.Lstat(filepath.Join(rootPath, symlink))
+	if assert.NoError(t, err) {
+		assert.EqualValues(t, filepath.Base(symlink), f.Name())
+		assert.Equal(t, fs.ModeSymlink, f.Mode().Type())
+	}
+	// cannot symlink out of root
+	assert.Error(t, client.Symlink(outbound, symlink))
+}
+
+func TestChrootReadDir(t *testing.T) {
+	rootPath, client := getChrootTuple(t)
+	defer client.Close(t)
+	const (
+		FILE_CNT = 10
+		MAX_SIZE = 1024
+	)
+	rand.Seed(time.Now().UnixNano())
+	files := make(map[string]int)
```

Comment:
Since this is a map of filenames to filesizes, I think this is more clearly named `fileSizes`. I woudl expect a map named `files` to be of type `map[string]*os.File`